CHANGELOG.markdown

```diff
@@ -1,3 +1,9 @@
+## ???
+
+* Improve type inference for `makeBaseFunctor` in recent GHC versions, and
+  remove its accidental ability to accept things like lists of declaration
```

Comment:
it wasn't accidental. It was very much deliberate, though maybe not really useful.

Overlapping instance is ugly. List of names instance isn't really useful, I'd just remove it. It's as "useful" as list of declarations quotes. Someone could written

```haskell
makeBaseFunctor
  [ [d|  instance Functor f => Recursive (Rose f a) |]
  , [d|  instance Functor f => Recursive (SomethingElse f a) |]
  , ...
  ]
```

FWIW, this is breaking change.

Reply:
I guess anyone can write their own traversal if they want to use a list of names, though it's not very compact. I'll take it out if you'd prefer.

Reply:
Combinators like

```haskell
makeBaseFunctors :: MakeBaseFunctor a => [a] -> DecsQ
makeBaseFunctorsWith :: MakeBaseFunctor a => BaseRules -> [a] -> DecsQ
```

can be added. No need to be clever and guess stuff. Overlapping instances is almost always a wrong solution.

Reply:
Sure. I figured it'd be fine since the whole `MakeBaseFunctor` class is a hack anyway. It'd be cleaner to just have

```haskell
makeBaseFunctorName :: Name -> DecsQ
makeBaseFunctorsNames :: [Name] -> DecsQ
makeBaseFunctorsInstances :: DecsQ -> DecsQ
```
----

By the way: For more recent `template-haskell` versions, it would theoretically be nicer to have

```haskell
makeBaseFunctorName :: Quasi m => Name -> m [Dec]
makeBaseFunctorsNames :: Quasi m => [Name] -> m [Dec]
makeBaseFunctorsInstances :: Quasi m => m [Dec] -> m [Dec]
```

to avoid needing to call `runQ` on `makeBaseFunctorName` and `makeBaseFunctorsNames`, and to allow `makeBaseFunctorsInstances` to use any special facilities of the ambient `Quasi` to generate the instance declarations. I doubt it's worth the compatibility headache for now.